### PR TITLE
Enable publishNotReadyAddresses on pd/tikv/tidb headless service

### LINKED ISSUE #1041: Enable publishNotReadyAddresses on headless service for stable DNS records
## Feature Request

**Is your feature request related to a problem? Please describe:**

Pod needs to become ready in order to have a record unless `publishNotReadyAddresses=True` is set on the Service.

**Describe the feature you'd like:**

Stable network DNS records regardless of pod state.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/

### PR DESCRIPTION
**Pull request by cofyc**
### What problem does this PR solve? 

fixes #1041

### Check List 

Tests 

 - Unit test
 - E2E test
 - Stability test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has Helm charts change
 - Has Go code change
 - Has CI related scripts change
 - Has documents change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note

 ```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -144,7 +144,7 @@ lint:
 tidy:
 	@echo "go mod tidy"
 	go mod tidy
-	git diff --quiet go.mod go.sum
+	git diff -U --exit-code go.mod go.sum
```
Comment (cofyc) [inline Makefile]: more useful output

File: hack/pin-deps.sh [hunk 2]
```diff
@@ -24,10 +27,16 @@ else
...
+)
 for repo in ${STAGING_REPOS[@]}; do
 	edit_args+=(-replace $repo=$repo@kubernetes-$VERSION)
 done
 
 go mod edit ${edit_args[@]}
+# workaround for https://github.com/golang/go/issues/33008
+# go mod tidy does not always remove unncessary liens from go.sum. For now we
+# can remove it first and populate again.
+rm go.sum
```
Comment (cofyc) [inline hack/pin-deps.sh]: regenerate `go.sum` file to remove unnecessary liens

File: hack/pin-deps.sh [hunk 3]
```diff
@@ -24,10 +27,16 @@ else
...
+edit_args=(
+    -fmt
+)
 for repo in ${STAGING_REPOS[@]}; do
 	edit_args+=(-replace $repo=$repo@kubernetes-$VERSION)
 done
 
 go mod edit ${edit_args[@]}
+# workaround for https://github.com/golang/go/issues/33008
+# go mod tidy does not always remove unncessary liens from go.sum. For now we
```
Comment (DanielZhangQD) [inline hack/pin-deps.sh]: liens -> lines?

Comment (cofyc) [reply hack/pin-deps.sh]: fixed, thanks!

**Comment by onlymellb** [comment]
Can we add a switch to control whether it is turned on? Maybe in some scenarios, we need to resolve to the available endpoints.

**Comment by cofyc** [comment]
> Can we add a switch to control whether it is turned on? Maybe in some scenarios, we need to resolve to the available endpoints.

The purpose of headless service is to provide stable network identifiers for members of StatefulSet.  It does not have `ClusterIP`. We created another service when we want to access available (ready) endpoints, see [`getNewPDServiceForTidbCluster`](https://github.com/pingcap/tidb-operator/blob/4a4c9825ca6cece90a5c35e67197e723d8a4589c/pkg/manager/member/pd_member_manager.go#L343-L370).

**Pull request merged**